### ISSUE TITLE
How to get/use RouteContext in TenantResolver

### ISSUE THREAD

**Issue by romario-mota** [quarkusio/quarkus#38125]
Hello,

I'm trying to use the RoutingContext in the TenantResolver, to set the tenantId. In my CustomSecurityAugmentor, I make a call to the database with panache to define the user's permissions.
I realized that if I don't make this database call, I can use the RoutingContext in the TenantResolver. However, if I do it it throws "jakarta.enterprise.inject.IllegalProductException: Normal scoped producer method may not return null".
The opposite is also true. If I do not use the RoutingContext in my TenantResolver, the database call in CustomSecurityAugmentor occurs successfully.

I need to use the RoutingContext in the TenantResolver to obtain the tenantId that is in the Request, and I also need to search for permissions in the database.

How do I make this work?

My Code:

**Tenant Resolver**

`
@PersistenceUnitExtension
@RequestScoped
public class CustomTenantResolver implements TenantResolver {

	public static final String DEFAULT_TENANT_ID = "00000000-0000-0000-0000-000000000000";

	@Inject
	RoutingContext context;

	@Override
	public String getDefaultTenantId() {
		return DEFAULT_TENANT_ID;
	}

	@Override
	public String resolveTenantId() {
		String tenant = context.get("tenant-id");
        String tenantId = tenant != null ? tenant : getDefaultTenantId();
		System.out.println("Tenant: "+tenantId);
		return tenantId;
	}
}`

**CustomSecurityAugmentor**

`
@ApplicationScoped
public class CustomSecurityAugmentor implements SecurityIdentityAugmentor {

	@Inject
	Instance<CustomSecurityIdentitySupplier> identitySupplierInstance;

	@Override
	public Uni<SecurityIdentity> augment(SecurityIdentity identity, AuthenticationRequestContext context) {
		// Hibernate ORM is blocking
		CustomSecurityIdentitySupplier identitySupplier = identitySupplierInstance.get();
		identitySupplier.setIdentity(identity);
		return context.runBlocking(identitySupplier);
	}
}

`

`@Dependent
public class CustomSecurityIdentitySupplier implements Supplier<SecurityIdentity> {

	@ConfigProperty(name = "quarkus.http.root-path")
	String rootPath;
	private SecurityIdentity identity;
	@Inject
	private PermissaoRepository permissaoRepository;

	public void setIdentity(SecurityIdentity identity) {
		this.identity = identity;
	}

	@ActivateRequestContext
	@Override
	public SecurityIdentity get() {

		if (identity.isAnonymous())
			return identity;
		String idUsuario = identity.getPrincipal().getName();
		System.out.println(String.format("UserId: %s", Converter.toString(idUsuario)));
		List<Permissao> permissoes = permissaoRepository.obtemPermissoesUsuarios(Converter.toUUID(idUsuario));
		// create a new builder and copy principal, attributes, credentials and roles
		// from the original identity
		QuarkusSecurityIdentity.Builder builder = QuarkusSecurityIdentity.builder(identity);
		// add custom role source here
		for(Permissao p : permissoes)
			builder.addRole(p.getId());
		return builder.build();

	}
}`

Thanks.

**Comment by geoand** [comment]
If there is no HTTP request active, you get the problem you see

**Comment by yrodiere** [comment]
Yeah I doubt `@ActivateRequestContext` will cut it in this case. If you call `CustomSecurityIdentitySupplier#get` outside of a request, e.g. in a background thread pool or a test that didn't send a request, then you'll get a request context, but the routing context won't work, since how could it: there is no routing being performed, not URL, etc.

If you want to make this work even when not using requests, you'll probably have to check whether the routing context is available before using it in your `TenantResolver`, and have a fallback strategy to determine the tenant ID when there is no routing context. Injecting `Instance<RoutingContext>` and calling `Instance#isAvailable` might help, though I'm not entirely sure.

Or you could try setting up some dummy `RoutingContext` on top of using `@ActivateRequestContext`, but I don't now how to do that.

**Comment by geoand** [comment]
What does your code example where an HTTP request is handled and Hibernate used look like?

**Comment by romario-mota** [comment]
In addition to the code I already included above, there is also a Filter where I place the tenant in the RoutingContext:

```java
@PreMatching
@Provider
public class TenantRequestFilter implements ContainerRequestFilter {

	@Inject
	RoutingContext context;
	@Inject
	JsonWebToken token;

	@Override
	public void filter(ContainerRequestContext requestContext) throws IOException {
		context.put("tenant-id", token.getClaim("tenant-id"));
	}
}
```

and the database call:

```java
@ApplicationScoped
public class PermissaoRepository implements PanacheRepository<Permissao> {

    public List<Permissao> obtemPermissoesUsuarios(UUID id) {
        return  this.list("select p from Permissao p join p.usuarios u where u.idExterno = ?1", id);
    }

}
```

**Comment by geoand** [comment]
So just to make sure we are all on the same page, can you describe at a high level what exactly your end goal is?

**Comment by romario-mota** [comment]
Hi,

After some tests, I verified that when executing the permissions query with panache in CustomSecurityIdentitySupplier, hibernate tries to resolve the Tenant, and at this moment it is not possible to access the RoutingContext in the TenantResolver. I suspect it's because the authentication isn't finished yet.

The only way I managed to make it work was by running the permissions query in CustomSecurityIdentitySupplier through raw JDBC, without Hibernate. This way, he doesn't try to resolve the tenant.

```java

    @Inject
    DataSource dataSource;

    public boolean usuarioTemPermissao(UUID idUsuario, String idPermissao) {
        String sql = ".....";
        try(Connection connection = dataSource.getConnection(); PreparedStatement ps = connection.prepareStatement(sql);) {
            ps.setObject(1, idUsuario);
            ps.setString(2, idPermissao);
            ResultSet rs = ps.executeQuery(); rs.next();
            return rs.getInt(1) > 0;
        }catch (Exception erro){
            ...
        }
    }

```

Thank you for the support.
